stl/inc/__msvc_int128.hpp

```diff
@@ -0,0 +1,1155 @@
+        }
+        return *this;
+    }
+    constexpr _Uint128 operator--(int) noexcept {
+        auto _Tmp = *this;
+        --*this;
+        return _Tmp;
+    }
+
+    constexpr _Uint128 operator+() const noexcept {
+        return *this;
+    }
+
+    constexpr _Uint128 operator-() const noexcept {
+        return _Uint128{} - *this;
+    }
+
+    constexpr _Uint128 operator~() const noexcept {
+        return _Uint128{~_Word[0], ~_Word[1]};
+    }
+
+    _NODISCARD friend constexpr _Uint128 operator+(const _Uint128& _Left, const _Uint128& _Right) noexcept {
+        _Uint128 _Result;
+        const auto _Carry = _AddCarry64(0, _Left._Word[0], _Right._Word[0], _Result._Word[0]);
+        _AddCarry64(_Carry, _Left._Word[1], _Right._Word[1], _Result._Word[1]);
+        return _Result;
+    }
+
+    constexpr _Uint128& operator+=(const _Uint128& _That) noexcept {
+        const auto _Carry = _AddCarry64(0, _Word[0], _That._Word[0], _Word[0]);
+        _AddCarry64(_Carry, _Word[1], _That._Word[1], _Word[1]);
+        return *this;
+    }
+    template <integral _Ty>
+    friend constexpr _Ty& operator+=(_Ty& _Left, const _Uint128& _Right) noexcept {
+        _Left = static_cast<_Ty>(_Uint128{_Left} + _Right);
+        return _Left;
+    }
+
+    _NODISCARD friend constexpr _Uint128 operator-(const _Uint128& _Left, const _Uint128& _Right) noexcept {
+        _Uint128 _Result;
+        const auto _Borrow = _SubBorrow64(0, _Left._Word[0], _Right._Word[0], _Result._Word[0]);
+        _SubBorrow64(_Borrow, _Left._Word[1], _Right._Word[1], _Result._Word[1]);
+        return _Result;
+    }
+
+    constexpr _Uint128& operator-=(const _Uint128& _That) noexcept {
+        const auto _Borrow = _SubBorrow64(0, _Word[0], _That._Word[0], _Word[0]);
+        _SubBorrow64(_Borrow, _Word[1], _That._Word[1], _Word[1]);
+        return *this;
+    }
+    template <integral _Ty>
+    friend constexpr _Ty& operator-=(_Ty& _Left, const _Uint128& _Right) noexcept {
+        _Left = static_cast<_Ty>(_Uint128{_Left} - _Right);
+        return _Left;
+    }
+
+    _NODISCARD friend constexpr _Uint128 operator*(const _Uint128& _Left, const _Uint128& _Right) noexcept {
+        _Uint128 _Result;
+        _Result._Word[0] = _UMul128(_Left._Word[0], _Right._Word[0], _Result._Word[1]);
+        _Result._Word[1] += _Left._Word[0] * _Right._Word[1];
+        _Result._Word[1] += _Left._Word[1] * _Right._Word[0];
+        return _Result;
+    }
+
+    constexpr _Uint128& operator*=(const _Uint128& _That) noexcept {
+        *this = *this * _That;
+        return *this;
+    }
+    template <integral _Ty>
+    friend constexpr _Ty& operator*=(_Ty& _Left, const _Uint128& _Right) noexcept {
+        _Left = static_cast<_Ty>(_Uint128{_Left} * _Right);
+        return _Left;
+    }
+
+    _NODISCARD friend constexpr _Uint128 operator/(const _Uint128& _Num, const uint32_t _Den) noexcept {
+        _Uint128 _Result;
+        _Result._Word[1] = _Num._Word[1] / _Den;

```

Comment:
Suboptimal due to DevCom-1544675
Also, there are `_div128` / `_udiv128`

Reply:
> Suboptimal due to [DevCom-1544675](https://developercommunity.visualstudio.com/t/1544675)

Filed #2520 to apply workarounds in a followup.

> Also, there are `_div128` / `_udiv128`

Yes! But not on Clang. For clang (and non-X64) this function is far better than using the `_udiv128` implementation in `_UDiv128`. I'll wrap this (and the equivalent `% uint32_t`) in some conditional compilation goop so it isn't used on MSVC/x64 where the `uint64_t` operations are clearly better.